File: ArgentSetupUi/Logging/SetupLog.cs
using System;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Threading;

namespace Argent.SetupUi.Logging
{
    /// <summary>
    /// Append-only setup log matching legacy SETUP_LOG.TXT / SAMPLE.TXT format.
    /// </summary>
    public sealed class SetupLog
    {
        private static readonly SetupLog Instance = new SetupLog();
        private readonly object _lock = new object();
        private int _sequence;
        private bool _headerWritten;
        private readonly string _machine;
        private readonly string _user;
        private readonly int _processId;

        public static SetupLog Current => Instance;

        public string LogFilePath { get; }

        private SetupLog()
        {
            _machine = Environment.MachineName;
            _user = (Environment.UserName ?? string.Empty).ToUpperInvariant();
            _processId = Process.GetCurrentProcess().Id;

            var logsDir = Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                "Argent", "Setup", "LOGS");
            Directory.CreateDirectory(logsDir);
            LogFilePath = Path.Combine(logsDir, "SETUP_LOG.TXT");
        }

        public void EnsureHeader()
        {
            lock (_lock)
            {
                if (_headerWritten || File.Exists(LogFilePath))
                {
                    _headerWritten = true;
                    return;
                }

                var header =
                    Environment.NewLine + Environment.NewLine +
                    "Argent Job Scheduler 10.0-2401-64W-A Setup Log File" + Environment.NewLine +
                    "Copyright (c) 2024 Argent Software" + Environment.NewLine +
                    Environment.NewLine +
                    "This product is protected by one or more of the following U.S. Patents:" + Environment.NewLine +
                    Environment.NewLine +
                    "6483813; 511167; 511346; 530335; 543551; 553142; 553143; 553144;" + Environment.NewLine +
                    " 553626; 553627; 553628; 553629; 553630; 553631; 553635; 553636;" + Environment.NewLine +
                    " 553637; 562835; 562836; 562837." + Environment.NewLine +
                    Environment.NewLine +
                    "Please contact Support at Support@Argent.com or help.Argent.com." + Environment.NewLine;

                File.AppendAllText(LogFilePath, header);
                _headerWritten = true;
            }
        }

        public void LogSessionStart()
        {
            EnsureHeader();
            Log(string.Empty);
            LogLabel("Machine", _machine);
            Log(string.Empty);
            Log(string.Empty);
            LogLabel("User", Environment.UserName ?? string.Empty);
            Log(string.Empty);
            LogLabel("Product", "Argent Job Scheduler");
            Log(string.Empty);
            LogLabel("Version", "10.0-2401-A");
            Log(string.Empty);
            LogLabel("Program", Environment.Is64BitProcess ? "64bit" : "32bit");
            Log(string.Empty);
            Log("Setup UI session started");
        }

        public void Log(string message)
        {
            lock (_lock)
            {
                EnsureHeader();
                File.AppendAllText(LogFilePath, FormatLine(message) + Environment.NewLine);
            }
        }

        public void LogLabel(string label, string value)
        {
            Log(string.Format(CultureInfo.InvariantCulture, "{0,-20}{1}", label + ":", value ?? string.Empty));
        }

        public void LogQuoted(string label, string value)
        {
            Log(string.Format(CultureInfo.InvariantCulture, "{0} '{1}'", label, value ?? string.Empty));
        }

        private string FormatLine(string message)
        {
            var now = DateTime.Now;
            var timestamp = now.ToString("dd MMM yyyy HH:mm:ss.fff", CultureInfo.InvariantCulture);
            var threadId = Thread.CurrentThread.ManagedThreadId;
            _sequence++;
            return string.Format(
                CultureInfo.InvariantCulture,
                "{0} {1} {2} ({3:D5}-{4:D5}) {5:D5} {6}",
                timestamp,
                _machine,
                _user,
                _processId,
                threadId,
                _sequence,
                message ?? string.Empty);
        }
    }
}

File: desktop/setupLog.js
const fs = require("fs");
const path = require("path");
const os = require("os");

const MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];

let sequence = 0;
let headerWritten = false;

function logDir() {
  const base = process.env.LOCALAPPDATA || path.join(os.homedir(), "AppData", "Local");
  return path.join(base, "Argent", "Setup", "LOGS");
}

function logFilePath() {
  return path.join(logDir(), "SETUP_LOG.TXT");
}

function formatTimestamp(d) {
  const pad2 = (n) => String(n).padStart(2, "0");
  const pad3 = (n) => String(n).padStart(3, "0");
  return `${pad2(d.getDate())} ${MONTHS[d.getMonth()]} ${d.getFullYear()} ${pad2(d.getHours())}:${pad2(
    d.getMinutes()
  )}:${pad2(d.getSeconds())}.${pad3(d.getMilliseconds())}`;
}

function formatLine(message) {
  const d = new Date();
  const machine = os.hostname();
  const user = (os.userInfo().username || "").toUpperCase();
  const pid = process.pid;
  const tid = 0;
  sequence += 1;
  return `${formatTimestamp(d)} ${machine} ${user} (${String(pid).padStart(5, "0")}-${String(tid).padStart(
    5,
    "0"
  )}) ${String(sequence).padStart(5, "0")} ${message ?? ""}`;
}

function ensureHeader() {
  const file = logFilePath();
  if (headerWritten || fs.existsSync(file)) {
    headerWritten = true;
    return;
  }
  fs.mkdirSync(logDir(), { recursive: true });
  const header = `

Argent Job Scheduler 10.0-2401-64W-A Setup Log File
Copyright (c) 2024 Argent Software

This product is protected by one or more of the following U.S. Patents:

6483813; 511167; 511346; 530335; 543551; 553142; 553143; 553144;
 553626; 553627; 553628; 553629; 553630; 553631; 553635; 553636;
 553637; 562835; 562836; 562837.

Please contact Support at Support@Argent.com or help.Argent.com.
`;
  fs.appendFileSync(file, header, "utf8");
  headerWritten = true;
}

function logLabel(label, value) {
  const padded = `${label}:`.padEnd(20, " ");
  logEvent(`${padded}${value ?? ""}`);
}

function logEvent(message) {
  fs.mkdirSync(logDir(), { recursive: true });
  ensureHeader();
  fs.appendFileSync(logFilePath(), `${formatLine(message)}\n`, "utf8");
}

function logSessionStart() {
  ensureHeader();
  logEvent("");
  logLabel("Machine", os.hostname());
  logEvent("");
  logEvent("");
  logLabel("User", os.userInfo().username || "");
  logEvent("");
  logLabel("Product", "Argent Job Scheduler");
  logEvent("");
  logLabel("Version", "10.0-2401-A");
  logEvent("");
  logLabel("Program", process.arch.includes("64") ? "64bit" : "32bit");
  logEvent("");
  logEvent("Setup UI session started");
}

module.exports = { logEvent, logLabel, logSessionStart, logFilePath };
